src/wallet/sendTransaction.ts

```diff
@@ -0,0 +1,353 @@
+
+    this.mineTransaction.on('unexpected-error', (message) => {
+      this.emit('send-error', message);
+    })
+
+    this.mineTransaction.on('success', (data) => {
+      this.transaction!.parents = data.parents;
+      this.transaction!.timestamp = data.timestamp;
+      this.transaction!.nonce = data.nonce;
+      this.transaction!.weight = data.weight;
+      this.emit('mine-tx-ended', data);
+    })
+
+    if (newOptions.startMiningTx) {
+      this.mineTransaction.start();
+    }
+
+    return this.mineTransaction;
+  }
+
+  /**
+   * Create and send a tx proposal to wallet service
+   * Expects this.transaction to be prepared, signed and mined
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  async handleSendTxProposal() {
+    if (this.transaction === null) {
+      throw new WalletError('Can\'t push transaction if it\'s null.');
+    }
+    this.emit('send-tx-start', this.transaction);
+    const txHex = this.transaction.toHex();
+
+    try {
+      const responseData = await walletApi.createTxProposal(this.wallet, txHex);
+      const txProposalId = responseData.txProposalId;
+      const sendData = await walletApi.updateTxProposal(this.wallet, txProposalId, txHex);
+      this.transaction.updateHash();
+      this.emit('send-tx-success', this.transaction);
+    } catch (err) {
+      if (err instanceof WalletRequestError) {
+        this.emit('send-error', 'Error sending tx proposal.');
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Run sendTransaction from mining, i.e. expects this.transaction to be prepared and signed
+   * then it will mine and handle tx proposal
+   *
+   * 'until' parameter can be 'mine-tx', in order to only mine the transaction without propagating
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  runFromMining(until = null) {
+    try {
+      this.mineTx();
+      if (until === 'mine-tx') {
+        return;
+      }
+
+      this.once('mine-tx-ended', (data) => {
+        this.handleSendTxProposal();
+      });
+    } catch (err) {
+      if (err instanceof WalletError) {
+        this.emit('send-error', err);
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Run sendTransaction from preparing, i.e. prepare, sign, mine and send the tx
+   *
+   * 'until' parameter can be 'prepare-tx' (it will stop before signing the tx), 'sign-tx' (it will stop before mining the tx),
+   * or 'mine-tx' (it will stop before send tx proposal, i.e. propagating the tx)
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  async run(until = null) {
+    try {
+      const preparedData = await this.prepareTx();
+      if (until === 'prepare-tx') {
+        return;
+      }
+
+      this.signTx(preparedData.utxosAddressPath);
+      if (until === 'sign-tx') {
+        return;
+      }
+
+      this.runFromMining(until);
+    } catch (err) {
+      if (err instanceof WalletError) {
+        this.emit('send-error', err);
```

Comment:
Why are we omitting this exception?

Reply:
The same as in this comment here: https://github.com/HathorNetwork/hathor-wallet-lib/pull/221#discussion_r687262959